Optimized all refactorings, fixed bugs analysed by SonarQube/SonarCloud

LHommeDeBat opened the pull request:
**Bugs/Code-Smells:**
- fixed bugs analysed by SonarCloud
- fixed some code smells analysed by SonarCloud
- CheckRepository method within the APIGrabber also uses the GithubToken now. It works without it but the max amount of requests per hour is too low without authorization and can cause problems while creating new configurations

**ReorderModifier:**
- Will not create PR or push to git if all modifiers are correct order. Will respond to user with an error message, telling him that everything is in correct order already
- Newer versions of JavaParser have issues with old code (Modifier code changed and now has some (known) problems with the LexicalPreservingPrinter). Rewritten this refactoring for that purpose to refactor it manually like with the "RemoveCommentedOutCode"

**RenameMethod:**
- Rewritten refactoring class to drastically improve performance on bigger projects by using the build in Javaparser functions to find super/sub classes
- Added more checks  (like is newMethodName = oldMethodName -> throw error)
- If method to rename might be an overridden method by a external project -> Renaming will be done but the commit message will contain a warning. This is neccessary so that projects that depend on many external classes can be refactored properly and won't throw errors almost every time

**RemoveParameter:**
- Rewritten the same way as "RenameMethod"
- Refactoring can only be successfull if no other method that is in some way related to the "toRefactorMethod" uses the "to be removed parameter"
- Added multiple checks to ensure proper error messages/bot responses on comments
- Added function for the refactoring to work with sonarcloud.

**AddOverrideAnnotation:**
- Bot will not add an Override annotation to a method that already has one. Will instead throw an error and respond to the PR-Comment accordingly

MarvinWyrich commented:
Thank you very much for the improvements made.

> Newer versions of JavaParser have issues with old code (Modifier code changed and now has some (known) problems with the LexicalPreservingPrinter). Rewritten this refactoring for that purpose to refactor it manually like with the "RemoveCommentedOutCode"

Should be fixed with [javaparser/PR#1993](https://github.com/javaparser/javaparser/pull/1993) if it's the same problem we already know. We can then use the LexicalPreservationPrinter again to avoid error-prone manual modifications of the source code. The manual modification for the RemoveCommentedOutCode refactoring is also just a temporary solution.

I will review the remaining changes soon. I hope that the Javaparser pull request will be merged into the master soon and that we can benefit from the fix.

MarvinWyrich commented on src/main/java/de/refactoringBot/refactoring/supportedRefactorings/RenameMethod.java:
```diff
@@ -126,43 +148,189 @@ public String performRefactoring(BotIssue issue, GitConfiguration gitConfig) thr
+					ancestors = getAllAncestors(currentClass.resolve());
+					hasExternalDep = true;
+				} catch (Exception e) {
+					continue;
```
At least log the exception message at debug level

LHommeDeBat replied:
Added logs at debug level.

LHommeDeBat commented on src/main/java/de/refactoringBot/refactoring/supportedRefactorings/RemoveMethodParameter.java:
```diff
@@ -122,341 +149,296 @@ public String performRefactoring(BotIssue issue, GitConfiguration gitConfig) thr
-	private ParserRefactoringCollection createRefactoringObjects(ParserRefactoringCollection allRefactorings,
-			String methodSignature, String classSignature, String javaFile, List<String> allJavaFiles)
-			throws FileNotFoundException {
+	public List<String> findJavaRoots(List<String> allJavaFiles, String repoFolder) throws FileNotFoundException {
```
I placed the method inside the new RefactoringHelper-Class. Not sure where else to put it. It is tied to refactoring and dynamic since package names can change, so I can't save those to our Config-Files that we store in our DB.

LHommeDeBat replied:
I also placed all methods there that are used in multiple refactorings to reduce duplicate code.

LHommeDeBat commented on src/main/java/de/refactoringBot/refactoring/supportedRefactorings/RemoveMethodParameter.java:
```diff
@@ -103,14 +102,42 @@ public String performRefactoring(BotIssue issue, GitConfiguration gitConfig) thr
+						ancestors = getAllAncestors(currentClass.resolve());
+						refactoring.setWarning(
+								" Refactored classes might extend/implement external project! Check if overriden method was NOT renamed!");
+					} catch (Exception e) {
```
Added more specific exceptions and modified error messages at some places

MarvinWyrich commented:
@LHommeDeBat javaparser has been updated. Can you check if the issue with the preservingPrinter has been resolved and then use it (again) in the refactoring classes?

LHommeDeBat commented:
> @LHommeDeBat javaparser has been updated. Can you check if the issue with the preservingPrinter has been resolved and then use it (again) in the refactoring classes?

Still getting "java.lang.UnsupportedOperationException: kept child(ClassOrInterfaceType) vs "String"   <89>" when trying to set correctly ordered "Modifier(s)" to our "FieldDeclaration". But for some reason only in some instances. It works for example with my Github project but does not work with our Calculator project, even though I'm using the same Field Declaration. (public final static int MULTIPLIER = 10;)

MarvinWyrich commented:
Thank you. Except for one thing, it looks good.

In the RemoveParameter refactoring there should be a check for method overloading with different numbers of parameters (-> no refactoring then). Otherwise [it will result in a compilation error](https://github.com/MarvinWyrich/Bot-Playground/pull/5/files).

LHommeDeBat commented:
> Thank you. Except for one thing, it looks good.
> 
> In the RemoveParameter refactoring there should be a check for method overloading with different numbers of parameters (-> no refactoring then). Otherwise [it will result in a compilation error](https://github.com/MarvinWyrich/Bot-Playground/pull/5/files).

Should I not refactor anything in that case or just leave that specific method unrefactored? (since there is already a method declaration we need)

MarvinWyrich commented:
> Should I not refactor anything in that case or just leave that specific method inside that specific class unrefactored? (since the class already has a method declaration we need and the method calls will still work without the param since the overloaded method exists)

Good point. I'm honestly not sure what the best approach would be. You would have to leave *all* methods in this class unrefactored to avoid an infinite recursion. And we can't be sure that the already existing method (with less parameters) doesn't do any semantically different things, so we can't just let all callers call this different method. (In the referenced PR example it is not a problem as one method just calls the other one but this does not have the be the case.)

I therefore tend to play it safe and to not refactor anything if removing a parameter would create a method signature that already exists. What do you think?

LHommeDeBat commented:
> > Should I not refactor anything in that case or just leave that specific method inside that specific class unrefactored? (since the class already has a method declaration we need and the method calls will still work without the param since the overloaded method exists)
> 
> Good point. I'm honestly not sure what the best approach would be. You would have to leave _all_ methods in this class unrefactored to avoid an infinite recursion. And we can't be sure that the already existing method (with less parameters) doesn't do any semantically different things, so we can't just let all callers call this different method. (In the referenced PR example it is not a problem as one method just calls the other one but this does not have the be the case.)
> 
> I therefore tend to play it safe and to not refactor anything if removing a parameter would create a method signature that already exists. What do you think?
> 
> /edit: in addition, we wouldn't be able to fix the code smell either.

You're right. The other method could be doing something entirely different. Added a check that will throw an error and not refactor anything. Code smell will stay that way but I don't see what else to do. the user has to rename the method first and then remove the parameter (which he can with our bot).

MarvinWyrich commented:
@LHommeDeBat We should also remove the parameter description from the javadoc (if present). I have added some unit tests that should pass once implemented.

MarvinWyrich commented:
> Refactoring can only be successfull if no other method that is in some way related to the "toRefactorMethod" uses the "to be removed parameter"

I think that the current implementation is a little bit too strict here. We should not compare the parameter names in the called methods and always throw an exception, because the parameter names in two methods can be the same without having anything to do with each other.

Find an example in the [failing testcase that I've just added](https://github.com/Refactoring-Bot/Refactoring-Bot/pull/17/commits/14ffb1288a17f24d46e14ae4246d5d1d3a2cd624). I'm afraid that this test case is not a special case and we should support the refactoring of this method.

LHommeDeBat commented:
> > Refactoring can only be successfull if no other method that is in some way related to the "toRefactorMethod" uses the "to be removed parameter"
> 
> I think that the current implementation is a little bit too strict here. We should not compare the parameter names in the called methods and always throw an exception, because the parameter names in two methods can be the same without having anything to do with each other.
> 
> Find an example in the [failing testcase that I've just added](https://github.com/Refactoring-Bot/Refactoring-Bot/pull/17/commits/14ffb1288a17f24d46e14ae4246d5d1d3a2cd624). I'm afraid that this test case is not a special case and we should support the refactoring of 

> > Refactoring can only be successfull if no other method that is in some way related to the "toRefactorMethod" uses the "to be removed parameter"
> 
> I think that the current implementation is a little bit too strict here. We should not compare the parameter names in the called methods and always throw an exception, because the parameter names in two methods can be the same without having anything to do with each other.
> 
> Find an example in the [failing testcase that I've just added](https://github.com/Refactoring-Bot/Refactoring-Bot/pull/17/commits/14ffb1288a17f24d46e14ae4246d5d1d3a2cd624). I'm afraid that this test case is not a special case and we should support the refactoring of this method.

I will look into the "removing from javadoc" thing. I have a problem with the testcase "testRemoveUnusedParameter2". It throws an error, saying that "a" is used in an completely different method, but if I refactore the method via. a GitHub comment, it works just fine. With my current implementaton, it should not look in the other method at all since these are not on the same AST-Tree. (Working example: https://github.com/LHommeDeBat/TestPullRequest/pull/398/files). Can you specify what exactly the failing case is?

LHommeDeBat commented:
I added the param-removal from the JavaDoc of a method if the param is removed from the method signature. There is however an strange formation of the edited Javadoc (Example: https://github.com/LHommeDeBat/TestPullRequest/pull/406/files).

I also fixed the one test-error. There is however still an "Failure", also inside the "testRemoveUnusedParameter2"-Testcase. It says "expected<2> but was <3>". It seems it expects that the method calls of our refactored method have 2 Parmeters instead of 3 but the method that was refactored went from 1 parameter to 0. Here it also seems that the testcase is expecting a different method to be refactored.

LHommeDeBat commented:
> > Refactoring can only be successfull if no other method that is in some way related to the "toRefactorMethod" uses the "to be removed parameter"
> 
> I think that the current implementation is a little bit too strict here. We should not compare the parameter names in the called methods and always throw an exception, because the parameter names in two methods can be the same without having anything to do with each other.
> 

I think there is an misunderstanding. I only throw an error if there is a method inside a super/subclass with the same signature which is using the parameter. According to "https://refactoring.guru/remove-parameter" you should not remove a parameter if it is used in a super/subclass since it can either lead to many syntax errors (Example: interface and two classes that implement it. One class uses the param, other doesn't. If I remove it from one class, i have to remove it from the interface and then I would have to remove it from the other class, which I can't because it is used there). Even if we find a case where it doesn't lead to syntax errors, and remove the param from the super but not the subclass, we would make an overriden method not overriden anymore, which I think we shoud not, what do you think?
